operations/agent_monitor.go

```diff
@@ -0,0 +1,427 @@
+			SetErrorSender(level.Error, grip.GetSender()).
+			ApplyFromOpts(opts).
+			Background(true)
+		if err = cmd.Run(ctx); err != nil {
+			return true, errors.Wrap(err, "failed to create process")
+		}
+
+		jasperIDs := cmd.GetProcIDs()
+
+		// The command should spawn exactly 1 process.
+		if len(jasperIDs) != 1 {
+			catcher := grip.NewBasicCatcher()
+
+			for _, jasperID := range jasperIDs {
+				proc, err = m.jasperClient.Get(ctx, jasperID)
+				if err != nil {
+					catcher.Add(errors.Wrap(err, "could not find erroneous process"))
+					continue
+				}
+				catcher.Add(proc.Signal(ctx, syscall.SIGKILL))
+			}
+
+			return true, errors.Wrapf(catcher.Resolve(), "monitor should spawn exactly 1 process (the agent), but found %d", len(jasperIDs))
+		}
+
+		proc, err = m.jasperClient.Get(ctx, jasperIDs[0])
+		if err != nil {
+			return true, errors.Wrap(err, "could not find agent process")
+		}
+
+		return false, nil
+	}, retry); err != nil {
+		return nil, errors.Wrapf(err, "failed to start agent process '%v'", agentCmdArgs)
+	}
+	return proc, nil
+}
+
+// waitUntilComplete waits until the given process to complete or the context
+// times out. maxDelay determines the time between checks on the process.
+func waitUntilComplete(ctx context.Context, proc jasper.Process, maxDelay time.Duration) (exitCode int, err error) {
+	timer := time.NewTimer(0)
+	defer timer.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return proc.Wait(ctx)
+		case <-timer.C:
+			if proc.Complete(ctx) {
+				return proc.Wait(ctx)
+			}
+			timer.Reset(maxDelay)
+		}
+	}
+}
+
+// runAgent starts the agent with the necessary args and waits for it to
+// terminate.
+func runAgent(ctx context.Context, m *monitor, retry *retryArgs) error {
+	proc, err := createAgentProcess(ctx, m, retry)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create agent process")
+	}
+
+	exitCode, err := waitUntilComplete(ctx, proc, defaultRetryArgs().maxDelay)
+
+	return errors.Wrapf(err, "agent exited with code %d", exitCode)
+}
+
+// runMonitor runs the monitor loop. It fetches the agent, starts it, and
+// repeats when the agent terminates.
+func runMonitor(ctx context.Context, m *monitor) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			// The evergreen agent runs using a separate binary from the monitor
+			// to allow the agent to be updated.
+			if err := fetchClient(ctx, m, defaultRetryArgs()); err != nil {
+				grip.Error(message.WrapError(err, message.Fields{
+					"message":     "could not fetch client",
+					"client_url":  m.clientURL,
+					"client_path": m.clientPath,
+				}))
+				continue
+			}
+
+			if _, err := os.Stat(m.clientPath); os.IsNotExist(err) {
+				grip.Error(errors.Wrapf(err, "could not stat client '%s'", m.clientPath))
+				continue
+			}
+
+			if err := runAgent(ctx, m, defaultRetryArgs()); err != nil {
+				grip.Error(errors.Wrap(err, "error occurred while running the agent"))
+			}
+		}
```

Comment:
should we log occasionally? grip.InfoWhen(sometimes.Fifth(), message.Fields{})

should there be some kind of sleep/iteration between rounds just to avoid spinning too hot, particularly during deploys

Reply:
I added a 30 second sleep after the first loop, but can change it if it's too low/high.